testament/testament.nim

```diff
@@ -128,6 +128,18 @@ proc execCmdEx2(command: string, args: openArray[string]; workingDir, input: str
   verboseCmd(result.cmdLine)
   var p = startProcess(command, workingDir = workingDir, args = args,
                        options = {poStdErrToStdOut, poUsePath})
+
+  if timeout > 0.0:
+    let first = epochTime()
+
+    while running(p):
+      if epochTime() - first > timeout:
+        terminate(p)
+        close(p)
+        return
+      else:
+        cpuRelax()
```

Comment:
Too dangerous: You start the process, then wait for it to finish and then capture the output stream. But the process could block just because the output stream wasn't read. I think.

Reply:
I'm not sure about the situation the child process will wait for the parent process to read the result for testing environment. I hope someone can clarify it a bit or provide a simple example.

Reply:
```nim
discard """
  timeout: 5.0
"""
# in tests/misc/t17407.nim
for i in 0..<100000:
  echo i
```

indeed, this test runs in ~0.5 seconds with `nim r tests/misc/t17407.nim` but would fail with reTimeout with your PR with `testament r tests/misc/t17407.nim` (in fact, no matter what timeout you'd put) because of what @Araq is describing.

the fix is to read the output line by line, avoiding a deadlock, and add the timeout check after each call to readLine.

you can add `tests/misc/t17407.nim` as a regression test